### ISSUE TITLE
Restore CRD panic if the CRD without structural schema

### ISSUE THREAD

**Issue by jenting** [vmware-tanzu/velero#2249]
**What steps did you take and what happened:**

Install velero by [helm chart](https://github.com/vmware-tanzu/helm-charts), backup and restore the CRD, bump out errors when describing the restore name.
`rpc error: code = Aborted desc = plugin panicked: runtime error: invalid memory address or nil pointer dereference`

**What did you expect to happen:**

Restore CRD without error.

**The output of the following commands will help us better understand what's going on**:

I think the main problem comes with this line https://github.com/vmware-tanzu/velero/blob/5b1280c/pkg/restore/crd_v1_preserve_unknown_fields_action.go#L79

Not every CRD includes OpenAPIV3Schema, such as https://github.com/vmware-tanzu/helm-charts/blob/e32ce47/charts/velero/crds/backups.yaml and https://github.com/dexidp/dex/blob/34dcf6c/scripts/manifests/crds/authcodes.yaml.

I think the correct solution to fix CRD is when backing up, to back up its original CRD apiVersion, not server-preferred apiVersion which is `v1` version in k8s 1.16+ cluster.
Then at restoring, velero restores the original CRD apiVersion.
My 2 cents.

The problem is how to backup the original CRD apiVersion :thinking: ? Since after k8s 1.16+ cluster, via webhook conversion, all the CRD apiVersion converts from `v1beta1` to `v1`.

---

* `kubectl logs deployment/velero -n velero`
* `velero backup describe <backupname>` or `kubectl get backup/<backupname> -n velero -o yaml`
* `velero backup logs <backupname>`
* `velero restore describe <restorename>` or `kubectl get restore/<restorename> -n velero -o yaml`
```
Errors:
  Velero:     <none>
  Cluster:  error preparing customresourcedefinitions.apiextensions.k8s.io/authcodes.dex.coreos.com: rpc error: code = Aborted desc = plugin panicked: runtime error: invalid memory address or nil pointer dereference
            error preparing customresourcedefinitions.apiextensions.k8s.io/authrequests.dex.coreos.com: rpc error: code = Aborted desc = plugin panicked: runtime error: invalid memory address or nil pointer dereference
            error preparing customresourcedefinitions.apiextensions.k8s.io/backups.velero.io: rpc error: code = Aborted desc = plugin panicked: runtime error: invalid memory address or nil pointer dereference
            error preparing customresourcedefinitions.apiextensions.k8s.io/backupstoragelocations.velero.io: rpc error: code = Aborted desc = plugin panicked: runtime error: invalid memory address or nil pointer dereference
            error preparing customresourcedefinitions.apiextensions.k8s.io/connectors.dex.coreos.com: rpc error: code = Aborted desc = plugin panicked: runtime error: invalid memory address or nil pointer dereference
            error preparing customresourcedefinitions.apiextensions.k8s.io/deletebackuprequests.velero.io: rpc error: code = Aborted desc = plugin panicked: runtime error: invalid memory address or nil pointer dereference
            error preparing customresourcedefinitions.apiextensions.k8s.io/downloadrequests.velero.io: rpc error: code = Aborted desc = plugin panicked: runtime error: invalid memory address or nil pointer dereference
            error preparing customresourcedefinitions.apiextensions.k8s.io/oauth2clients.dex.coreos.com: rpc error: code = Aborted desc = plugin panicked: runtime error: invalid memory address or nil pointer dereference
            error preparing customresourcedefinitions.apiextensions.k8s.io/offlinesessionses.dex.coreos.com: rpc error: code = Aborted desc = plugin panicked: runtime error: invalid memory address or nil pointer dereference
            error preparing customresourcedefinitions.apiextensions.k8s.io/passwords.dex.coreos.com: rpc error: code = Aborted desc = plugin panicked: runtime error: invalid memory address or nil pointer dereference
            error preparing customresourcedefinitions.apiextensions.k8s.io/podvolumebackups.velero.io: rpc error: code = Aborted desc = plugin panicked: runtime error: invalid memory address or nil pointer dereference
            error preparing customresourcedefinitions.apiextensions.k8s.io/podvolumerestores.velero.io: rpc error: code = Aborted desc = plugin panicked: runtime error: invalid memory address or nil pointer dereference
            error preparing customresourcedefinitions.apiextensions.k8s.io/refreshtokens.dex.coreos.com: rpc error: code = Aborted desc = plugin panicked: runtime error: invalid memory address or nil pointer dereference
            error preparing customresourcedefinitions.apiextensions.k8s.io/resticrepositories.velero.io: rpc error: code = Aborted desc = plugin panicked: runtime error: invalid memory address or nil pointer dereference
            error preparing customresourcedefinitions.apiextensions.k8s.io/restores.velero.io: rpc error: code = Aborted desc = plugin panicked: runtime error: invalid memory address or nil pointer dereference
            error preparing customresourcedefinitions.apiextensions.k8s.io/schedules.velero.io: rpc error: code = Aborted desc = plugin panicked: runtime error: invalid memory address or nil pointer dereference
            error preparing customresourcedefinitions.apiextensions.k8s.io/serverstatusrequests.velero.io: rpc error: code = Aborted desc = plugin panicked: runtime error: invalid memory address or nil pointer dereference
            error preparing customresourcedefinitions.apiextensions.k8s.io/signingkeies.dex.coreos.com: rpc error: code = Aborted desc = plugin panicked: runtime error: invalid memory address or nil pointer dereference
            error preparing customresourcedefinitions.apiextensions.k8s.io/volumesnapshotlocations.velero.io: rpc error: code = Aborted desc = plugin panicked: runtime error: invalid memory address or nil pointer dereference
  Namespaces: <none>

Backup:  backup2slave

Namespaces:
  Included:  *
  Excluded:  <none>

Resources:
  Included:        *
  Excluded:        nodes, events, events.events.k8s.io, backups.velero.io, restores.velero.io, resticrepositories.velero.io
  Cluster-scoped:  auto

Namespace mappings:  <none>

Label selector:  <none>

Restore PVs:  auto
```
* `velero restore logs <restorename>`

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Velero version (use `velero version`): 
```
Client:
	Version: v1.2.0
	Git commit: 5d008491bbf681658d3e372da1a9d3a21ca4c03c
Server:
	Version: v1.3.0-beta.1`
```
- Velero features (use `velero client config get features`): 
```
features: <NOT SET>
```
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.2", GitCommit:"c97fe5036ef3df2967d086711e6c0c405941e14b", GitTreeState:"clean", BuildDate:"2019-10-15T19:18:23Z", GoVersion:"go1.12.10", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.2", GitCommit:"c97fe5036ef3df2967d086711e6c0c405941e14b", GitTreeState:"clean", BuildDate:"2019-10-28T12:00:00Z", GoVersion:"go1.12.12", Compiler:"gc", Platform:"linux/amd64"}
```
- Kubernetes installer & version:
- Cloud provider or hardware configuration:
- OS (e.g. from `/etc/os-release`):

**Comment by carlisia** [comment]
For next week: c/c @nrb + @skriss

**Comment by nrb** [comment]
@jenting Can you provide the specific CRD that you has issues with?

**Comment by jenting** [comment]
> @jenting Can you provide the specific CRD that you has issues with?

- https://github.com/vmware-tanzu/helm-charts/blob/e32ce47/charts/velero/crds/backups.yaml
- https://github.com/dexidp/dex/blob/34dcf6c/scripts/manifests/crds/authcodes.yaml

**Comment by nrb** [comment]
@jenting Thanks, I'll test that today. Thank you for trying out the beta!

**Comment by nrb** [comment]
@jenting To make sure I'm testing the same thing you did, what Kubernetes version(s) did you use? Did you back up and restore from the same version?

**Comment by nrb** [comment]
So I'm able to reproduce this behavior, and I can prevent the panic with this change:

```diff
+                       // If the Schema is nil, there are no nested fields to set, so skip over it for this version.
+                       if v.Schema == nil {
+                               continue
+                       }
+
                        // Use the address, since the XPreserveUnknownFields value is undefined or true (false is not allowed)
                        preserve := true
                        v.Schema.OpenAPIV3Schema.XPreserveUnknownFields = &preserve
```

However, that then leads to more issues. We're backing up the Helm-installed CRDs as `v1`, but they don't have any schema associated, which is invalid for v1.

We may actually want to add more smarts to this via a plugin to decide whether or not to use v1 or v1beta1, as they use different rules. One boolean field is easy enough, but we can't recreate a schema for someone else's CRDs.

**Comment by nrb** [comment]
An example:

```
error restoring customresourcedefinitions.apiextensions.k8s.io/schedules.velero.io: CustomResourceDefinition.apiextensions.k8s.io "schedules.velero.io" is invalid: spec.versions[0].schema.openAPIV3Schema: Required value: schemas are required
error restoring customresourcedefinitions.apiextensions.k8s.io/serverstatusrequests.velero.io: CustomResourceDefinition.apiextensions.k8s.io "serverstatusrequests.velero.io" is invalid: spec.versions[0].schema.openAPIV3Schema: Required value: schemas are required
error restoring customresourcedefinitions.apiextensions.k8s.io/volumesnapshotlocations.velero.io: CustomResourceDefinition.apiextensions.k8s.io "volumesnapshotlocations.velero.io" is invalid: spec.versions[0].schema.openAPIV3Schema: Required value: schemas are required
```

**Comment by nrb** [comment]
I'm going to experiment with maybe flagging v1beta1 CRDs that don't have a schema defined at backup time to not use the preferred version.

On Kubernetes 1.16, they clearly work when submitted to the v1beta1 endpoint and are made available for reads via the v1 endpoint just fine, but the problem manifests when Velero backs them up from the v1 endpoint and resubmits them to back to v1.

For Velero, we definitely need to update our CRDs for the Helm chart sooner rather than later.

**Comment by skriss** [comment]
So I guess ideally:
- we'd be able to reliably detect which CRDs could not be created as `v1`
- we'd back those up & restore them as `v1beta1`

On the first point - do we know if missing schemas is the only condition that would prevent creating as `v1`? (I know about the PreserveUnknownFields=true issue too, but we have already added logic to convert that)

I'm not exactly sure how we'd accomplish the second point yet - I'm envisioning some hacky combination of backup/restore item actions. Need to think on it some more.  If we always back up the `v1` representation, is it possible for us to write an item action that "down-converts" it back to `v1beta1`? I'm not sure how easy/possible this is off the top of my head.

Alternately, a blunter approach - we could add some hardcoding somewhere to always back up `v1beta1` CRDs, for now, until we can sort this out better.

**Comment by jenting** [comment]
@nrb 
Yes, I back up and restore on the same Kubersion version `1.16.2`.

I had filed a [PR](https://github.com/vmware-tanzu/helm-charts/pull/39) to update Velero helm-charts already, PTAL :bow:

**Comment by jenting** [comment]
> So I guess ideally:
> 
> * we'd be able to reliably detect which CRDs could not be created as `v1`
> * we'd back those up & restore them as `v1beta1`
> 
> On the first point - do we know if missing schemas is the only condition that would prevent creating as `v1`? (I know about the PreserveUnknownFields=true issue too, but we have already added logic to convert that)

structural schema is optional in `v1beta1`.

> 
> I'm not exactly sure how we'd accomplish the second point yet - I'm envisioning some hacky combination of backup/restore item actions. Need to think on it some more. If we always back up the `v1` representation, is it possible for us to write an item action that "down-converts" it back to `v1beta1`? I'm not sure how easy/possible this is off the top of my head.
> 
> Alternately, a blunter approach - we could add some hardcoding somewhere to always back up `v1beta1` CRDs, for now, until we can sort this out better.

I would vote for always back up `v1beta1` CRDs if no better approach on how to know the CRD original version.

**Comment by skriss** [comment]
> structural schema is optional in `v1beta1`.

Yep, so if we see a CRD that *doesn't* have a structural schema, then we know it was created as `v1beta1`. I'm just wondering if there are any other such fields that would signal that a CRD was created as `v1beta1`. `PreserveUnknownFields=true` would be an example of these other scenarios that I'm looking for.

**Comment by nrb** [comment]
>I'm just wondering if there are any other such fields that would signal that a CRD was created as v1beta1

I'll put this out as a question in the community meeting today, and also do some research on my own. So far `(Schema=nil and PreserveUnknownFields=true)` are the only heuristic I've found for a CRD that is only restorable as `v1beta1`, but it wouldn't surprise me if I'd missed some other case here.

**Comment by skriss** [comment]
xref https://github.com/kubernetes/kubernetes/issues/87231 - previous issue @jenting logged upstreams. You can see what liggitt has to say.

**Comment by nrb** [comment]
After reviewing the Kubernets API docs and Steve's linked issue, I've created a PR linked to this issue with 2 possible solutions that appear to work - one sets the apiVersion field to v1beta1 on backup, and the other keeps the CRD as v1 but populates the necessary fields with empty schema structs.

I've listed some pros and cons I see for both approaches in the PR as comments, so I'd appreciate thoughts on those. Both implementations seem fairly minor, so I don't think that will be a blocker in either case.

Also, I think having some defensive logic around nil pointers on the restore side will make sense regardless of the fix on the back up side.
